src/plugins/dist_matrixops/all_gather.cpp

```diff
@@ -0,0 +1,383 @@
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter, "all_gather::all_gather2d",
+                generate_error_message(
+                    "the operand has incompatible dimensionalities"));
+        }
+
+        switch (extract_common_type(arr))
+        {
+        case node_data_type_bool:
+            return all_gather2d(
+                extract_boolean_value_strict(std::move(arr), name_, codename_),
+                std::move(locs));
+
+        case node_data_type_int64:
+            return all_gather2d(
+                extract_integer_value_strict(std::move(arr), name_, codename_),
+                std::move(locs));
+
+        case node_data_type_unknown:
+            return all_gather2d(
+                extract_numeric_value(std::move(arr), name_, codename_),
+                std::move(locs));
+
+        case node_data_type_double:
+            return all_gather2d(
+                extract_numeric_value_strict(std::move(arr), name_, codename_),
+                std::move(locs));
+
+        default:
+            break;
+        }
+
+        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+            "dist_matrixops::primitives::all_gather::all_gather2d",
+            generate_error_message(
+                "the all_gather_d primitive requires for all arguments to "
+                "be numeric data types"));
+
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    hpx::future<execution_tree::primitive_argument_type> all_gather::eval(
+        execution_tree::primitive_arguments_type const& operands,
+        execution_tree::primitive_arguments_type const& args,
+        execution_tree::eval_context ctx) const
+    {
+        if (operands.empty() || operands.size() > 1)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "all_gather::eval",
+                generate_error_message(
+                    "the all_gather primitive requires"
+                    "1 operands"));
+        }
+
+        if (!valid(operands[0]))
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "all_gather::eval",
+                generate_error_message(
+                    "the all_gather primitive requires the first argument"
+                    "given by the operands array is valid"));
+        }
+
+        auto this_ = this->shared_from_this();
+        return hpx::dataflow(hpx::launch::sync,
+            hpx::util::unwrapping(
+                [this_ = std::move(this_)](
+                    execution_tree::primitive_arguments_type&& args)
+                    -> execution_tree::primitive_argument_type
+                    {
+                        using namespace execution_tree;
+
+                        switch (extract_numeric_value_dimension(
+                            std::move(args[0]), this_->name_, this_->codename_))
+                        {
+
+                            case 1:
+                                HPX_FALLTHROUGH;
```

Comment:
If I understand it correctly, it should work for 1d arrays, too. I didn't find any test for that, would you please add some?

Reply:
I think I should delete 1d case here. Because I only consider 2d case. The reason why I added 1d here is that  I assumed a matrix which has only one row or one column is 1d. But it should still be 2d, right?

Reply:
yeah, I believe that's a 2d, too.

Reply:
OK. Will delete that 1d case